### ISSUE TITLE
Label block devices with duplicate ids

### ISSUE THREAD

**Issue by kinvaris** [openvstorage/framework#1231]
Related to: http://tickets.openvstorage.com/browse/OVSSUP-125

On a customer environment we have seperated storage nodes that have 1 NVMe each. This NVMe is mapped under `/dev/nvme0n1`. But in the OVS GUI we see all the NVMe's being mapped under 1 node.

PS: Maybe its a good idea to test this also with FusionIO drives.

![](https://s12.postimg.org/6e1973pcd/ovs.png)

![](https://s18.postimg.org/bar6925kp/Selection_015.png)

![](https://s18.postimg.org/vw5y0yn5l/Selection_014.png)

![](https://s15.postimg.org/fogod8k6j/Selection_016.png)

**Comment by khenderick** [comment]
```
root@cpu-17:~# ls /dev/disk/by*/* -l | grep nvm
root@cpu-17:~# lsblk --nodeps -o name,serial
NAME    SERIAL
sda     E20150714150079
sdb     E20150714150074
sdc     E20150714150060
sdd     E20150714150059
sde     E20150714150053
sdf     E20150714150075
sdg     E20150714150056
sdh     E20150714150057
nvme0n1
root@cpu-17:~#
```

**Comment by khenderick** [comment]
To be checked whether the udev rules to give the nvme drives a unique identifier is correctly installed. We need a unique way of identifying disks. If not available, the above will be observed.

**Comment by kinvaris** [comment]
# Possible workaround

## Symptoms:

### The NVMe had no unique id and theirfor showed weird behavior on in the OVS GUI.

```
root@cpu-18:~# lsblk --nodeps -o name,serial
NAME    SERIAL
nvme0n1             <-- NO SERIAL

ls /dev/disk/by*/* -l | grep nvm            <-- NO OUTPUT
```

## Solution:

### Create a udev rule for the nvmes with the following content

* `vim /lib/udev/rules.d/99-ovs-nvme-naming.rules`

```
# nvme
KERNEL=="nvme[0-9]*", PROGRAM="/opt/scripts/nvme-serial.sh $env{DEVNAME}", ENV{ID_SERIAL}="%c"
KERNEL=="nvme[0-9]*", ENV{DEVTYPE}=="disk", ENV{ID_SERIAL}=="?*", SYMLINK+="disk/by-id/nvme-$env{ID_SERIAL}"
KERNEL=="nvme[0-9]*", ENV{DEVTYPE}=="partition", ENV{ID_SERIAL}=="?*", SYMLINK+="disk/by-id/nvme-$env{ID_SERIAL}-part%n"
```

## Perform the following steps:

* `apt-get update`
* `apt-get install nvme-cli`
* `mkdir -p /opt/scripts/`
* `vim /opt/scripts/nvme-serial.sh`

```
#!/bin/sh
nvme id-ctrl $1 | awk '/sn/ { printf("%s", $3) }'
```

* `chmod +x /opt/scripts/nvme-serial.sh`
* `udevadm control --reload-rules`
* `udevadm trigger`

## Result

```
root@cpu-17:~# ls /dev/disk/by*/* -l | grep nvm
lrwxrwxrwx 1 root root 13 Dec  1 15:58 /dev/disk/by-id/nvme-PHMD546300PE400FGN -> ../../nvme0n1
```

**Comment by kinvaris** [comment]
@khenderick: to avoid such cases the OVS gui could filter out disks who don't hold a valid serial number as @domsj suggested.

**Comment by khenderick** [comment]
It was decided before that the framework should show _all_ blockdevices. @wimpers, does this ticket change this decision?

**Comment by domsj** [comment]
@kinvaris I suggested to keep on showing the disk, but add some warning indicator, an explanation of what's wrong and how it can be fixed

**Comment by khenderick** [comment]
@domsj, don't agree with the "... and how it can be fixed" since we might not have a clue about that, except for "make sure the block device you're trying to use exposes some kind of serial-ish thingy to the kernel". We happen to have an udev rule and script that works for at least a few types of nvme drives, but ymmv.

**Comment by wimpers** [comment]
My viewpoint:
* don't block any block devices
* it is up to OPS to make sure the udev script is up to date/adjusted when we see problems.
* in case duplicate id are detected , replace the disk icon on the backend page by a red http://fontawesome.io/icon/question-circle/.
* update documentation that each block device should have a unique id.

**Comment by wimpers** [comment]
@khenderick what is the impact of this udev patch on the update? Any chance the update resets the changes?

**Comment by khenderick** [comment]
No, the framework doesn't care about the udev rules so it won't do any changes to it in any way. We don't know about the files, so we don't change the files.

**Comment by pploegaert** [comment]
Related to: http://tickets.openvstorage.com/browse/OVSSUP-125

**Comment by wimpers** [comment]
Added message to Essentials page in GitBook
